skorch/utils.py

```diff
@@ -135,6 +135,13 @@ def to_numpy(X):
     if isinstance(X, (tuple, list)):
         return type(X)(to_numpy(x) for x in X)
 
+    if hasattr(X, 'dataset') and hasattr(X, 'idx') and hasattr(X, 'indices'):
+        # We probably deal with a SliceDataset. Cannot use isinstance because we
+        # don't want to depend on helper.py.
+        if np.isscalar(X[0]):
+            return np.array([X[i] for i in range(len(X))])
+        return np.array([to_numpy(X[i]) for i in range(len(X))])
```

Comment:
Nit:

```suggestion
        Xs = [X[i] for i in range(len(X))]
        if np.isscalar(out[0]):
            return np.array(Xs)
        return np.array([to_numpy(x) for x in Xs])
```

Reply:
Re sklearn version: I was getting ahead of myself a bit :)

Re this suggestion: Could you explain why it is better? (Also, there is no `out` variable, is it referring to `X` or `Xs`?) It seems like with your suggestion, we iterate over `X` a second time if we don't deal with a scalar, is it not potentially more expensive?

Ideally, we could have a version that doesn't need to test the first element at all, but I'm not sure it's achievable.

Reply:
> I wanted to guard against the second X[0] access, because SliceDataset.transform can do some computation. I have not benchmarked it, but my sense is that the numpy conversation will take up majority of the compute compared to iterating X a second time.

Yeah, it's a difficult call. In the best case, a numpy conversion is very cheap because pytorch and numpy share a memory layout. But in the worst case, as you mentioned, we pay for one additional `dataset.__getitem__` call. I think the more cautious solution is to prevent the worst case, i.e. avoid an additional `__getitem__` but iterate through the list once more.

> We can have to_numpy support scalars and return np.asarray(scalar) if the input is is a scalar?

I think this could be a dangerous solution performance-wise, as we would be checking each item of the list. As is, we only check the first item and then assume that the subsequent items have the same type.

> Side note: A slightly nicer solution would be to define SliceDataset.__array__, so it knows how to turn itself into a ndarray.

I like that, I moved the conversion code into `SliceDataset.__array__`. Of course, performance is the same this way or that, but it looks cleaner.